File: core/meta_line_parser.py
from typing import Any, Dict, List
import logging


class MetaParseError(ValueError):
    """Raised when metadata parsing fails due to invalid format."""
    pass


def parse_meta_line(line: str) -> Dict[str, Any]:
    """
    Parse a metadata line from 'screenshotmanager' or 'lfs' format.

    Returns a dict with keys:
      - raw_text: str
      - type: 'screenshotmanager' or 'lfs'
      - index: int
      - author: {'id': str, 'displayName': str}
      - world: {'id': str, 'instanceId': str, 'name': str}
      - position: {'x': float, 'y': float, 'z': float}
      - rq: int
      - players: List[{'id': str, 'position': {'x':float,'y':float,'z':float}, 'displayName': str}]
    """
    # initialize defaults
    meta: Dict[str, Any] = {
        'raw_text': line,
        'type': None,
        'index': None,
        'author': {'id': '', 'displayName': ''},
        'world': {'id': '', 'instanceId': '', 'name': ''},
        'position': {'x': 0.0, 'y': 0.0, 'z': 0.0},
        'rq': 0,
        'players': [],
    }

    parts = [p.strip() for p in line.split('|')]
    if len(parts) < 2:
        raise MetaParseError(f"Line must have at least type and index: {line!r}")

    # type & index
    meta_type = parts[0]
    if meta_type not in ('screenshotmanager', 'lfs'):
        raise MetaParseError(f"Unknown meta type: {meta_type!r}")
    meta['type'] = meta_type

    try:
        meta['index'] = int(parts[1])
    except ValueError:
        raise MetaParseError(f"Invalid index (must be int): {parts[1]!r}")

    # mapping for field parsers
    field_parsers = {
        'author': _parse_author,
        'world': _parse_world,
        'pos': _parse_pos,
        'rq': _parse_request,
        'players': _parse_players,
    }

    # parse remaining segments
    for seg in parts[2:]:
        # Handle bare world segments (e.g. "wrld_xxx,instanceId,worldName")
        # from older screenshotmanager format that omits the "world:" prefix
        if seg.startswith('wrld_'):
            try:
                _parse_world(seg, meta)
            except MetaParseError as e:
                logging.warning(f"Failed to parse bare world segment: {e}")
            continue

        if ':' not in seg:
            continue
        key, val = seg.split(':', 1)
        parser = field_parsers.get(key)
        if parser:
            try:
                parser(val, meta)
            except MetaParseError as e:
                logging.warning(f"Failed to parse {key}: {e}")
                # Continue parsing other fields even if one fails
        # unknown keys are ignored

    return meta


def _parse_author(val: str, meta: Dict[str, Any]) -> None:
    try:
        auth_id, auth_name = val.split(',', 1)
    except ValueError:
        raise MetaParseError(f"Invalid author format: {val!r}")
    meta['author'] = {'id': auth_id, 'displayName': auth_name}


def _parse_world(val: str, meta: Dict[str, Any]) -> None:
    try:
        w_id, w_inst, w_name = val.split(',', 2)
    except ValueError:
        raise MetaParseError(f"Invalid world format: {val!r}")
    meta['world'] = {'id': w_id, 'instanceId': w_id + ':' + w_inst, 'name': w_name}


def _parse_pos(val: str, meta: Dict[str, Any]) -> None:
    coords = val.split(',')
    if len(coords) != 3:
        raise MetaParseError(f"Invalid pos format, needs 3 floats: {val!r}")
    try:
        x, y, z = (float(c) for c in coords)
    except ValueError:
        raise MetaParseError(f"Non-numeric pos values: {val!r}")
    meta['position'] = {'x': x, 'y': y, 'z': z}


def _parse_request(val: str, meta: Dict[str, Any]) -> None:
    try:
        meta['rq'] = int(val)
    except ValueError:
        raise MetaParseError(f"Invalid rq value (must be int): {val!r}")


def _parse_players(val: str, meta: Dict[str, Any]) -> None:
    """Parse players list, skipping malformed entries instead of failing completely."""
    players: List[Dict[str, Any]] = []
    for entry in val.split(';'):
        parts_p = entry.split(',', 4)
        if len(parts_p) != 5:
            logging.warning(f"Skipping malformed player entry: {entry!r}")
            continue
        pid, sx, sy, sz, name = parts_p
        try:
            px, py, pz = float(sx), float(sy), float(sz)
        except ValueError:
            logging.warning(f"Skipping player with invalid coords: {entry!r}")
            continue
        players.append({
            'id': pid,
            'position': {'x': px, 'y': py, 'z': pz},
            'displayName': name
        })
    meta['players'] = players


__all__ = ['MetaParseError', 'parse_meta_line']

File: src/HydrusTagger.Taggers/Vrchat/MetaLineParser.cs
using System.Globalization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace HydrusTagger.Taggers.Vrchat;

/// <summary>Raised when a metadata line cannot be parsed at all.</summary>
public sealed class MetaParseException : Exception
{
    public MetaParseException(string message) : base(message) { }
}

/// <summary>
/// Parses the legacy pipe-delimited "screenshotmanager" / "lfs" formats.
/// Port of <c>core/meta_line_parser.py</c>.
/// </summary>
/// <remarks>
/// Parsing is deliberately lenient below the top level: a malformed individual
/// field is logged and skipped rather than discarding the whole record, because
/// partial metadata still yields useful tags. Only a missing/unknown type or a
/// non-integer index is fatal.
/// </remarks>
public static class MetaLineParser
{
    private static readonly string[] KnownTypes = ["screenshotmanager", "lfs"];

    public static VrcMetadata Parse(string line, ILogger? log = null)
    {
        log ??= NullLogger.Instance;

        var meta = new VrcMetadata { RawText = line };

        var parts = line.Split('|').Select(p => p.Trim()).ToArray();
        if (parts.Length < 2)
        {
            throw new MetaParseException($"Line must have at least type and index: '{line}'");
        }

        var metaType = parts[0];
        if (!KnownTypes.Contains(metaType, StringComparer.Ordinal))
        {
            throw new MetaParseException($"Unknown meta type: '{metaType}'");
        }

        meta.Type = metaType;

        if (!int.TryParse(parts[1], NumberStyles.Integer, CultureInfo.InvariantCulture, out var index))
        {
            throw new MetaParseException($"Invalid index (must be int): '{parts[1]}'");
        }

        meta.Index = index;

        foreach (var seg in parts[2..])
        {
            // Older screenshotmanager output emits the world segment bare,
            // without the "world:" prefix.
            if (seg.StartsWith("wrld_", StringComparison.Ordinal))
            {
                TryField(log, "world", () => ParseWorld(seg, meta));
                continue;
            }

            var colon = seg.IndexOf(':', StringComparison.Ordinal);
            if (colon < 0)
            {
                continue;
            }

            var key = seg[..colon];
            var val = seg[(colon + 1)..];

            switch (key)
            {
                case "author":
                    TryField(log, key, () => ParseAuthor(val, meta));
                    break;
                case "world":
                    TryField(log, key, () => ParseWorld(val, meta));
                    break;
                case "pos":
                    TryField(log, key, () => ParsePosition(val, meta));
                    break;
                case "rq":
                    TryField(log, key, () => ParseRq(val, meta));
                    break;
                case "players":
                    TryField(log, key, () => ParsePlayers(val, meta, log));
                    break;
                default:
                    // Unknown keys are ignored.
                    break;
            }
        }

        return meta;
    }

    /// <summary>Attempt to parse one field, keeping the rest of the record on failure.</summary>
    private static void TryField(ILogger log, string field, Action parse)
    {
        try
        {
            parse();
        }
        catch (MetaParseException ex)
        {
            log.LogWarning("Failed to parse {Field}: {Message}", field, ex.Message);
        }
    }

    private static void ParseAuthor(string val, VrcMetadata meta)
    {
        var comma = val.IndexOf(',', StringComparison.Ordinal);
        if (comma < 0)
        {
            throw new MetaParseException($"Invalid author format: '{val}'");
        }

        meta.Author = new VrcAuthor
        {
            Id = val[..comma],
            DisplayName = val[(comma + 1)..],
        };
    }

    private static void ParseWorld(string val, VrcMetadata meta)
    {
        // Split into at most 3: the world name may itself contain commas.
        var first = val.IndexOf(',', StringComparison.Ordinal);
        if (first < 0)
        {
            throw new MetaParseException($"Invalid world format: '{val}'");
        }

        var second = val.IndexOf(',', first + 1);
        if (second < 0)
        {
            throw new MetaParseException($"Invalid world format: '{val}'");
        }

        var worldId = val[..first];
        var instance = val[(first + 1)..second];
        var name = val[(second + 1)..];

        meta.World = new VrcWorld
        {
            Id = worldId,
            // Matches the VRCX JSON convention, where instanceId is prefixed
            // with the world id.
            InstanceId = worldId + ":" + instance,
            Name = name,
        };
    }

    private static void ParsePosition(string val, VrcMetadata meta)
    {
        var coords = val.Split(',');
        if (coords.Length != 3)
        {
            throw new MetaParseException($"Invalid pos format, needs 3 floats: '{val}'");
        }

        if (!TryParseDouble(coords[0], out var x)
            || !TryParseDouble(coords[1], out var y)
            || !TryParseDouble(coords[2], out var z))
        {
            throw new MetaParseException($"Non-numeric pos values: '{val}'");
        }

        meta.Position = new VrcPosition { X = x, Y = y, Z = z };
    }

    private static void ParseRq(string val, VrcMetadata meta)
    {
        if (!int.TryParse(val, NumberStyles.Integer, CultureInfo.InvariantCulture, out var rq))
        {
            throw new MetaParseException($"Invalid rq value (must be int): '{val}'");
        }

        meta.Rq = rq;
    }

    private static void ParsePlayers(string val, VrcMetadata meta, ILogger log)
    {
        var players = new List<VrcPlayer>();

        foreach (var entry in val.Split(';'))
        {
            // Split into at most 5: a display name may contain commas.
            var fields = entry.Split(',', 5);
            if (fields.Length != 5)
            {
                log.LogWarning("Skipping malformed player entry: '{Entry}'", entry);
                continue;
            }

            if (!TryParseDouble(fields[1], out var px)
                || !TryParseDouble(fields[2], out var py)
                || !TryParseDouble(fields[3], out var pz))
            {
                log.LogWarning("Skipping player with invalid coords: '{Entry}'", entry);
                continue;
            }

            players.Add(new VrcPlayer
            {
                Id = fields[0],
                DisplayName = fields[4],
                Position = new VrcPosition { X = px, Y = py, Z = pz },
            });
        }

        meta.Players = players;
    }

    /// <summary>
    /// Float parsing matching Python's <c>float()</c>: invariant culture, and
    /// accepting the same specials ("inf", "nan").
    /// </summary>
    private static bool TryParseDouble(string s, out double value) =>
        double.TryParse(
            s,
            NumberStyles.Float | NumberStyles.AllowThousands,
            CultureInfo.InvariantCulture,
            out value);
}
